src/rsa/signing.rs

```diff
@@ -176,181 +201,194 @@ impl KeyPair {
-                // implicitly below, e.g. when `q_mod_p` is constructed.
-                let ((p, p_bits, dP), (q, q_bits, dQ, qInv)) =
-                    match q.verify_less_than(&p) {
-                    Ok(_)  => ((p, p_bits, dP), (q, q_bits, dQ, Some(qInv))),
-                    Err(_) => {
-                        // TODO: verify `q` and `qInv` are inverses (mod p).
-                        ((q, q_bits, dQ), (p, p_bits, dP, None))
-                    },
-                };
-
-                // XXX: Some steps are done out of order, but the NIST steps
-                // are worded in such a way that it is clear that NIST intends
-                // for them to be done in order. TODO: Does this matter at all?
-
-                // 6.4.1.4.3/6.4.1.2.1 - Step 1.
-
-                // Step 1.a is omitted, as explained above.
-
-                // Step 1.b is omitted per above. Instead, we check that the
-                // public modulus is 2048 to
-                // `PRIVATE_KEY_PUBLIC_MODULUS_MAX_BITS` bits. XXX: The maximum
-                // limit of 4096 bits is primarily due to lack of testing of
-                // larger key sizes; see, in particular,
-                // https://www.mail-archive.com/openssl-dev@openssl.org/msg44586.html
-                // and
-                // https://www.mail-archive.com/openssl-dev@openssl.org/msg44759.html.
-                // Also, this limit might help with memory management decisions
-                // later.
-
-                // Step 1.c. We validate e >= 65537.
-                let public_key = verification::Key::from_modulus_and_exponent(
-                    n, e, bits::BitLength::from_usize_bits(2048),
-                    super::PRIVATE_KEY_PUBLIC_MODULUS_MAX_BITS, 65537)?;
-
-                // 6.4.1.4.3 says to skip 6.4.1.2.1 Step 2.
-
-                // 6.4.1.4.3 Step 3.
-
-                // Step 3.a is done below, out of order.
-                // Step 3.b is unneeded since `n_bits` is derived here from `n`.
-
-                // 6.4.1.4.3 says to skip 6.4.1.2.1 Step 4. (We don't need to
-                // recover the prime factors since they are already given.)
-
-                // 6.4.1.4.3 - Step 5.
-
-                // Steps 5.a and 5.b are omitted, as explained above.
-
-                // Step 5.c.
-                //
-                // TODO: First, stop if `p < (√2) * 2**((nBits/2) - 1)`.
-                //
-                // Second, stop if `p > 2**(nBits/2) - 1`.
-                let half_n_bits = public_key.n_bits.half_rounded_up();
-                if p_bits != half_n_bits {
-                    return Err(error::Unspecified);
-                }
+    /// Constructs an RSA private key from components.
+    ///
+    /// Components are in big-endian, considered unsigned, and must not contain
+    /// leading zeros. Input aside this is just like `from_pkcs8()`. See the
+    /// documentation for `from_pkcs8()` for more details.
```

Comment:
Let's not repeat this part of the documentation about the form of each value in `KeyPairInput` here.